# DiabeticClassification
[https://archive.ics.uci.edu/ml/datasets/Diabetes+130-US+hospitals+for+years+1999-2008](https://archive.ics.uci.edu/ml/datasets/Diabetes+130-US+hospitals+for+years+1999-2008)

**CLASSIFICATION PROBLEM**

In [2]:
# Common Imports

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb
import seaborn as sns

from sklearn.model_selection import train_test_split

# No INFO logging cluttering the notebook plz
import logging
logging.getLogger().setLevel(logging.CRITICAL)

# I only want to see warning logs once
import warnings
warnings.filterwarnings(action='once')

# axis 0 is the axis that runs downward down the rows.
# axis 1 is the axis that runs horizontally across the columns
# the axis parameter controls which axis will be aggregated
# the axis parameter controls which axis will be collapsed

# Checking if we have the right version of sklearn. Should be > 0.21
import sklearn

sklearn.__version__


'0.20.3'

## Guidelines
1. You should select those techniques, methods or algorithms that allow you to develop each task according to the characteristics of the problem to be solved. 
Justify your choices and connect the obtained results among tasks.  
2. You should focus on the task “evaluate different algorithms”. We do not want you to apply machine learning algorithms as a black box, 
but we want you to explore their characteristics, parameters, and evaluation. Specific instructions for each approach will be given.

# Define Problem
   
### Basic Explanation
It is important to know if a patient will be readmitted in some hospital. The reason is that you can change the treatment, in order to avoid a readmission.

In this database, you have 3 different outputs:

* No readmission;
* A readmission in less than 30 days (this situation is not good, because maybe your treatment was not appropriate);
* A readmission in more than 30 days (this one is not so good as well the last one, however, the reason can be the state of the patient.
In this context, you can see different objective functions for the problem. You can try to figure out situations where the patient will not be readmitted, or if their are going to be readmitted in less than 30 days (because the problem can the the treatment), etc... Make your choice and let's help them creating new approaches for the problem.

### Content
"The data set represents 10 years (1999-2008) of clinical care at 130 US hospitals and integrated delivery networks. It includes over 50 features representing patient and hospital outcomes. Information was extracted from the database for encounters that satisfied the following criteria.

It is an inpatient encounter (a hospital admission).
It is a diabetic encounter, that is, one during which any kind of diabetes was entered to the system as a diagnosis.
The length of stay was at least 1 day and at most 14 days.
Laboratory tests were performed during the encounter.
Medications were administered during the encounter.
The data contains such attributes as patient number, race, gender, age, admission type, time in hospital, medical specialty of admitting physician, number of lab test performed, HbA1c test result, diagnosis, number of medication, diabetic medications, number of outpatient, inpatient, and emergency visits in the year before the hospitalization, etc."

#### features

* **Encounter ID**	Unique identifier of an encounter
* **Patient number**	Unique identifier of a patient
* **Race**	Values: Caucasian, Asian, African American, Hispanic, and other
* **Gender**	Values: male, female, and unknown/invalid
* **Age**	Grouped in 10-year intervals: 0, 10), 10, 20), …, 90, 100)
* **Weight**	Weight in pounds
* **Admission type**	Integer identifier corresponding to 9 distinct values, for example, emergency, urgent, elective, newborn, and not available
* **Discharge disposition**	Integer identifier corresponding to 29 distinct values, for example, discharged to home, expired, and not available
* **Admission source**	Integer identifier corresponding to 21 distinct values, for example, physician referral, emergency room, and transfer from a hospital
* **Time in hospital**	Integer number of days between admission and discharge
* **Payer code**	Integer identifier corresponding to 23 distinct values, for example, Blue Cross/Blue Shield, Medicare, and self-pay Medical
* **Medical specialty**	Integer identifier of a specialty of the admitting physician, corresponding to 84 distinct values, for example, cardiology, internal medicine, family/general practice, and surgeon
* **Number of lab procedures**	Number of lab tests performed during the encounter
* **Number of procedures** Numeric	Number of procedures (other than lab tests) performed during the encounter
* **Number of medications**	Number of distinct generic names administered during the encounter
* **Number of outpatient visits** Number of outpatient visits of the patient in the year preceding the encounter
* **Number of emergency visits**	Number of emergency visits of the patient in the year preceding the encounter
* **Number of inpatient visits**	Number of inpatient visits of the patient in the year preceding the encounter
* **Diagnosis 1**	The primary diagnosis (coded as first three digits of ICD9); 848 distinct values
* **Diagnosis 2**	Secondary diagnosis (coded as first three digits of ICD9); 923 distinct values
* **Diagnosis 3** Additional secondary diagnosis (coded as first three digits of ICD9); 954 distinct values
* **Number of diagnoses**	Number of diagnoses entered to the system 0%
* **Glucose serum test result**	Indicates the range of the result or if the test was not taken. Values: “>200,” “>300,” “normal,” and “none” if not measured
* **A1c test result**	Indicates the range of the result or if the test was not taken. Values: “>8” if the result was greater than 8%, “>7” if the result was greater than 7% but less than 8%, “normal” if the result was less than 7%, and “none” if not measured.
* **Change of medications**	Indicates if there was a change in diabetic medications (either dosage or generic name). Values: “change” and “no change”
* **Diabetes medications**	Indicates if there was any diabetic medication prescribed. Values: “yes” and “no”
* 24 features for medications	For the generic names: **metformin, repaglinide, nateglinide, chlorpropamide, glimepiride, acetohexamide, glipizide, glyburide, tolbutamide, pioglitazone, rosiglitazone, acarbose, miglitol, troglitazone, tolazamide, examide, sitagliptin, insulin, glyburide-metformin, glipizide-metformin, glimepiride- pioglitazone, metformin-rosiglitazone, and metformin- pioglitazone**, the feature indicates whether the drug was prescribed or there was a change in the dosage. Values: “up” if the dosage was increased during the encounter, “down” if the dosage was decreased, “steady” if the dosage did not change, and “no” if the drug was not prescribed
* **Readmitted**	Days to inpatient readmission. Values: “<30” if the patient was readmitted in less than 30 days, “>30” if the patient was readmitted in more than 30 days, and “No” for no record of readmission

# Analyze Data and Prepare Data

### Downloading data

In [3]:
!curl https://archive.ics.uci.edu/ml/machine-learning-databases/00296/dataset_diabetes.zip -o dataset_diabetes.zip
!unzip -o dataset_diabetes.zip

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 3268k  100 3268k    0     0  1392k      0  0:00:02  0:00:02 --:--:-- 1392k
Archive:  dataset_diabetes.zip
  inflating: dataset_diabetes/diabetic_data.csv  
  inflating: dataset_diabetes/IDs_mapping.csv  


In [4]:
local_data_path = './dataset_diabetes/diabetic_data.csv'

In [5]:
df = pd.read_csv(local_data_path, sep=',')
pd.set_option('display.max_columns', 500)     # Make sure we can see all of the columns
pd.set_option('display.max_rows', 10)         # Keep the output on one page
df

encounter_id  patient_nbr             race  gender      age weight  \
0            2278392      8222157        Caucasian  Female   [0-10)      ?   
1             149190     55629189        Caucasian  Female  [10-20)      ?   
2              64410     86047875  AfricanAmerican  Female  [20-30)      ?   
3             500364     82442376        Caucasian    Male  [30-40)      ?   
4              16680     42519267        Caucasian    Male  [40-50)      ?   
...              ...          ...              ...     ...      ...    ...   
101761     443847548    100162476  AfricanAmerican    Male  [70-80)      ?   
101762     443847782     74694222  AfricanAmerican  Female  [80-90)      ?   
101763     443854148     41088789        Caucasian    Male  [70-80)      ?   
101764     443857166     31693671        Caucasian  Female  [80-90)      ?   
101765     443867222    175429310        Caucasian    Male  [70-80)      ?   

        admission_type_id  discharge_disposition_id  admission_source_id  \
0                       6                        25                    1   
1                       1                         1                    7   
2                       1                         1                    7   
3                       1                         1                    7   
4                       1                         1                    7   
...                   ...                       ...                  ...   
101761                  1                         3                    7   
101762                  1                         4                    5   
101763                  1                         1                    7   
101764                  2                         3                    7   
101765                  1                         1                    7   

        time_in_hospital payer_code         medical_specialty  \
0                      1          ?  Pediatrics-Endocrinology   
1                      3          ?                         ?   
2                      2          ?                         ?   
3                      2          ?                         ?   
4                      1          ?                         ?   
...                  ...        ...                       ...   
101761                 3         MC                         ?   
101762                 5         MC                         ?   
101763                 1         MC                         ?   
101764                10         MC           Surgery-General   
101765                 6          ?                         ?   

        num_lab_procedures  num_procedures  num_medications  \
0                       41               0                1   
1                       59               0               18   
2                       11               5               13   
3                       44               1               16   
4                       51               0                8   
...                    ...             ...              ...   
101761                  51               0               16   
101762                  33               3               18   
101763                  53               0                9   
101764                  45               2               21   
101765                  13               3                3   

        number_outpatient  number_emergency  number_inpatient  diag_1  diag_2  \
0                       0                 0                 0  250.83       ?   
1                       0                 0                 0     276  250.01   
2                       2                 0                 1     648     250   
3                       0                 0                 0       8  250.43   
4                       0                 0                 0     197     157   
...                   ...               ...               ...     ...     ...   
101761                  0                 0          

In [6]:
# shuffling the dataframe for convenience
df = df.sample(frac=1).reset_index(drop=True)

In [7]:
# It's more readable to transpose the result of head()
df.head(10).T

0          1                2  \
encounter_id             110663580  173570754        154973820   
patient_nbr               19038663   87577434         88308783   
race                             ?  Caucasian  AfricanAmerican   
gender                      Female       Male           Female   
age                       [90-100)    [60-70)          [60-70)   
...                            ...        ...              ...   
metformin-rosiglitazone         No         No               No   
metformin-pioglitazone          No         No               No   
change                          Ch         No               No   
diabetesMed                    Yes        Yes              Yes   
readmitted                      NO        >30              <30   

                                       3          4                5  \
encounter_id                   155586054  166865976        169796190   
patient_nbr                     23493690   80618877         69405021   
race                     AfricanAmerican  Caucasian  AfricanAmerican   
gender                              Male       Male           Female   
age                              [70-80)    [80-90)          [40-50)   
...                                  ...        ...              ...   
metformin-rosiglitazone               No         No               No   
metformin-pioglitazone                No         No               No   
change                                No         Ch               No   
diabetesMed                          Yes        Yes              Yes   
readmitted                            NO         NO               NO   

                                 6          7          8          9  
encounter_id             189612462  111787350  164315106  167703396  
patient_nbr               58774716    8427483   89930862   23842557  
race                     Caucasian          ?  Caucasian  Caucasian  
gender                      Female     Female       Male       Male  
age                        [30-40)    [80-90)    [80-90)    [70-80)  
...                            ...        ...        ...        ...  
metformin-rosiglitazone         No         No         No         No  
metformin-pioglitazone          No         No         No         No  
change                          No         Ch         Ch         No  
diabetesMed                     No        Yes        Yes        Yes  
readmitted                     >30         NO         NO         NO  

[50 rows x 10 columns]

## Data exploration

### Drop duplicate columns.
Only keep one record per patient

In [8]:
df = df.drop_duplicates(subset= ['patient_nbr'], keep = 'first')

### Drop unnec columns
`patient_nbr` and `encounter_id` are columns that identify a person in the hospital. For the prediction of readmittance we do not want to keep them.

In [9]:
df.drop(['patient_nbr', 'encounter_id'], axis=1, inplace=True)

### Nan Values
* replace ? with np.Nan
* total number of cols missing
* for every row find the number of columns missing

In [10]:
df = df.replace('?',np.NaN)

In [11]:
# total number of cols missing
# resNans = df.apply(lambda x: x.isna().sum(),axis =1)
# print('number of nans: {}'.format(sum(1 for x in resNans if x > 0)))

In [12]:
def missing_values_summary(df):
        mis_val = df.isna().sum()
        mis_val_percent = 100 * df.isna().sum() / len(df)
        mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
        mis_val_table_ren_columns = mis_val_table.rename(
        columns = {0 : 'Missing Values', 1 : '% of Total Values'})
        mis_val_table_ren_columns = mis_val_table_ren_columns[
            mis_val_table_ren_columns.iloc[:,1] != 0].sort_values(
        '% of Total Values', ascending=False).round(1)
        print ("Your selected dataframe has " + str(df.shape[1]) + " columns.\n"      
            "There are " + str(mis_val_table_ren_columns.shape[0]) +
              " columns that have missing values.")
        return mis_val_table_ren_columns

source: [https://stackoverflow.com/questions/26266362/how-to-count-the-nan-values-in-a-column-in-pandas-dataframe#answer-39734251](https://stackoverflow.com/questions/26266362/how-to-count-the-nan-values-in-a-column-in-pandas-dataframe#answer-39734251)

In [13]:
missing_values_summary(df)

Your selected dataframe has 48 columns.
There are 7 columns that have missing values.


Missing Values  % of Total Values
weight                      68663               96.0
medical_specialty           34461               48.2
payer_code                  30556               42.7
race                         1914                2.7
diag_3                       1165                1.6
diag_2                        296                0.4
diag_1                         12                0.0

#### Dropping columns
Since weight, medical_speciality and payer_code contain so many missing values I will drop them. For the other missing values I will use a different strategy later on.

In [14]:
data = df.drop(['weight', 'medical_specialty', 'payer_code'], axis=1)
df = data

### Checking datatypes
As will be clear from the below, we have some numerical an categorical features.
<div class="alert alert-info"> 💡 <strong> Preprocessing </strong>
Numerical and  categorical features need different types of preprocessing strategies
</div>

In [15]:
df.columns.to_series().groupby(df.dtypes).groups

{dtype('int64'): Index(['admission_type_id', 'discharge_disposition_id', 'admission_source_id',
        'time_in_hospital', 'num_lab_procedures', 'num_procedures',
        'num_medications', 'number_outpatient', 'number_emergency',
        'number_inpatient', 'number_diagnoses'],
       dtype='object'),
 dtype('O'): Index(['race', 'gender', 'age', 'diag_1', 'diag_2', 'diag_3', 'max_glu_serum',
        'A1Cresult', 'metformin', 'repaglinide', 'nateglinide',
        'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide',
        'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose',
        'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton',
        'insulin', 'glyburide-metformin', 'glipizide-metformin',
        'glimepiride-pioglitazone', 'metformin-rosiglitazone',
        'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted'],
       dtype='object')}

### Exploring and prepping the target column
We have a multiclassification problem here.
I do not like the fact that we have "NO", "<30" and ">30" as possible outcomes.
I will update to simpler categories.

I decide to make it a binary classification predicint readmittance or not.
This means that both '>30' and '<30' will be the same readmittance category.

In [16]:
df.readmitted.value_counts()

NO     47787
>30    18660
<30     5071
Name: readmitted, dtype: int64

In [17]:
df.loc[df['readmitted'] == 'NO', 'readmitted'] = 'A'
df.loc[df['readmitted'] == '<30', 'readmitted'] = 'B'
df.loc[df['readmitted'] == '>30', 'readmitted'] = 'B'

#### Imbalance?
We see that in all of our records the outcomes for the target variables are not evenly distributed.
We have a lot more situations where there was  no readmittence compared to situation where there was.
Now we will downsample until each target outcome is represented by the same number of records.

In [18]:
df.readmitted.value_counts()

A    47787
B    23731
Name: readmitted, dtype: int64

In [19]:
label='readmitted'

g = df.groupby(label, group_keys=False)
balanced_df = pd.DataFrame(g.apply(lambda x: x.sample(g.size().min()))).reset_index(drop=True)

In [20]:
balanced_df.readmitted.value_counts()

B    23731
A    23731
Name: readmitted, dtype: int64

In [21]:
df = balanced_df

#### Data Volume
We have quite a big dataset (for my computer to handle at least).  
I will drop until we only have 3000 rows left. Otherwise it takes too long to evaluate all the different models time and time again.
I will randomly drop. Since we are randomly dropping in the balanced dataset this will not cause a significant imbalance to appear.

In [22]:
df = df.sample(frac=1).reset_index(drop=True)

In [23]:
data = df.iloc[:3000, :]
df = data

In [24]:
df.shape

(3000, 45)

In [25]:
df.head(5)

race  gender      age  admission_type_id  \
0        Caucasian  Female  [50-60)                  1   
1  AfricanAmerican    Male  [40-50)                  1   
2        Caucasian    Male  [30-40)                  3   
3        Caucasian  Female  [70-80)                  3   
4        Caucasian  Female  [50-60)                  1   

   discharge_disposition_id  admission_source_id  time_in_hospital  \
0                         3                    7                 9   
1                         1                    7                 1   
2                         1                    1                 5   
3                        18                    1                 1   
4                         6                    7                 3   

   num_lab_procedures  num_procedures  num_medications  number_outpatient  \
0                  65               4               28                  0   
1                  38               1                6                  1   
2                  42               3               26                  0   
3                   1               1                4                  0   
4                  42               1                9                  0   

   number_emergency  number_inpatient  diag_1 diag_2 diag_3  number_diagnoses  \
0                 0                 0     403    996    707                 9   
1                 0                 0     280    448    784                 5   
2                 0                 0     558    788    250                 4   
3                 0                 0     241    250    NaN                 2   
4                 0                 0  250.83    682    V45                 9   

  max_glu_serum A1Cresult metformin repaglinide nateglinide chlorpropamide  \
0          None      None        No          No          No             No   
1          None      None    Steady          No          No             No   
2          None      None        No          No          No             No   
3          None      None        No          No          No         Steady   
4          None      Norm        No          No          No             No   

  glimepiride acetohexamide glipizide glyburide tolbutamide pioglitazone  \
0          No            No        No        No          No           No   
1          No            No        No        No          No           No   
2          No            No        No        No          No           No   
3          No            No        No        No          No           No   
4          No            No        No        No          No           No   

  rosiglitazone acarbose miglitol troglitazone tolazamide examide citoglipton  \
0            No       No       No           No         No      No          No   
1            No       No       No           No         No      No          No   
2            No       No       No           No         No      No          No   
3            No       No       No           No         No      No          No   
4            No       No       No           No         No      No          No   

  insulin glyburide-metformin glipizide-metformin glimepiride-pioglitazone  \
0    Down                  No                  No                       No   
1      No                  No                  No                       No   
2    Down                  No                  No                       No   
3      No                  No                  No                       No   
4    Down                  No                  No                       No   

  metformin-rosiglitazone metformin-pioglitazone change diabetesMed readmitted  
0                      No                     No     Ch         Yes          B  
1                      No                     No     No         Yes          B  
2                      No                     No     Ch         Yes          B  
3                      No                     No     No         Yes          A  
4      

## Preprocessing

I will create a pipeline where I add the preprocessing steps and later evaluate the different models.


Source: [https://medium.com/vickdata/a-simple-guide-to-scikit-learn-pipelines-4ac0d974bdcf](https://medium.com/vickdata/a-simple-guide-to-scikit-learn-pipelines-4ac0d974bdcf)

In [26]:
X = df.iloc[:, df.columns != 'readmitted']
y = df.iloc[:, df.columns == 'readmitted']

In [27]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

#### Preprocessing strategy

As mentioned above depending on the type of data you need a different strategy to preprocess the data.
I will use the following:
<div class="alert alert-info"> 💡 <strong> Preprocessing</strong>

**Numerical**
* SimpleImputer: define the strategy that you will use to resolve missing data (nan). We will fill these with the mean value.
* StandardScaler: Standardize features by removing the mean and scaling to unit variance

**Categorical**
* SimpleImputer: define the strategy that you will use to resolve missing data (nan). We will fill these in with the constant 'incomplete'.
* OneHotEncoder: represent the categorical features as a 1-hot array. One hot encoding is a process by which categorical variables are converted into a form that could be provided to ML algorithms to do a better job in prediction. 


Please refers to [https://scikit-learn.org/](https://scikit-learn.org/) to find detailed information about these preprocessing strategies
</div>

In [28]:
numeric_features = X_train.select_dtypes(include=['int64', 'float64']).columns
print(numeric_features)

Index(['admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'num_lab_procedures', 'num_procedures',
       'num_medications', 'number_outpatient', 'number_emergency',
       'number_inpatient', 'number_diagnoses'],
      dtype='object')


In [29]:
categorical_features = X_train.select_dtypes(include=['object']).columns
print(categorical_features)

Index(['race', 'gender', 'age', 'diag_1', 'diag_2', 'diag_3', 'max_glu_serum',
       'A1Cresult', 'metformin', 'repaglinide', 'nateglinide',
       'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide',
       'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose',
       'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton',
       'insulin', 'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed'],
      dtype='object')


In [30]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder

numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
    ('scaler', StandardScaler())])

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='incomplete')),
    ('onehotencoder', OneHotEncoder(handle_unknown='ignore'))])

In [31]:
from sklearn.compose import ColumnTransformer

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)])

# Evaluate Algoritms

## Evaluation different models
We will be looping over different classification models.

<div class="alert alert-info"> 💡 <strong> Model evalution</strong>

Model evaluation strategy:
* loop over the different classification model
* score each model on accuracy
* calculate f1 score for each model
* retain the score of each model
* compare the score of each model
</div>

[https://medium.com/vickdata/a-simple-guide-to-scikit-learn-pipelines-4ac0d974bdcf#49aa](https://medium.com/vickdata/a-simple-guide-to-scikit-learn-pipelines-4ac0d974bdcf#49aa)

In [32]:
# ignore warnings for deprecated code in the libraries that we use
warnings.filterwarnings("ignore")

In [33]:
from sklearn.metrics import accuracy_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import classification_report
from sklearn.linear_model import LogisticRegression

classifiers = [
    KNeighborsClassifier(),
    DecisionTreeClassifier(),
    RandomForestClassifier(),
    MLPClassifier(),
    LogisticRegression()
    ]

In [34]:
target_names = ['A', 'B']
f1_scores = {}

for classifier in classifiers:
    print(type(classifier).__name__)
    
    pipe = Pipeline(steps=[('preprocessor', preprocessor),('classifier', classifier)])
    pipe.fit(X_train, y_train)   
    
    print("model accuracy score: %.3f" % pipe.score(X_test, y_test))
    
    y_pred = pipe.predict(X_test)
    report = classification_report(y_test, y_pred, target_names=target_names)
    print(report)
    
    report = classification_report(y_test, y_pred, target_names=target_names, output_dict=True)
    f1_scores[type(classifier).__name__] = report.get('weighted avg').get('f1-score')
    
    print('\n')

KNeighborsClassifier
model accuracy score: 0.553
              precision    recall  f1-score   support

           A       0.57      0.54      0.55       461
           B       0.54      0.57      0.55       439

   micro avg       0.55      0.55      0.55       900
   macro avg       0.55      0.55      0.55       900
weighted avg       0.55      0.55      0.55       900



DecisionTreeClassifier
model accuracy score: 0.558
              precision    recall  f1-score   support

           A       0.57      0.55      0.56       461
           B       0.55      0.56      0.55       439

   micro avg       0.56      0.56      0.56       900
   macro avg       0.56      0.56      0.56       900
weighted avg       0.56      0.56      0.56       900



RandomForestClassifier
model accuracy score: 0.552
              precision    recall  f1-score   support

           A       0.56      0.59      0.57       461
           B       0.54      0.51      0.53       439

   micro avg       0.55    

In [35]:
from pprint import pprint

pprint(f1_scores)

{'DecisionTreeClassifier': 0.5578389293859157,
 'KNeighborsClassifier': 0.5533068640668184,
 'LogisticRegression': 0.5751151734738692,
 'MLPClassifier': 0.5683008822734851,
 'RandomForestClassifier': 0.551648627742195}


# Improve Results

The performance is (I find it) poor. I will pick a model and try to tune it.
I decide to continue with RandomForest and try to further optimize it's result.

### Model tuning

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rf = Pipeline(steps=[('preprocessor', preprocessor),
                      ('classifier', RandomForestClassifier())])

In [ ]:
rf.fit(X_train, y_train)

In [ ]:
y_pred = rf.predict(X_test)

In [ ]:
# ignore warnings for deprecated code in the libraries that we use
warnings.filterwarnings("ignore")

We are going to tune the parameters of our model. These parameters are:
* n_estimators: actually the number of trees in our forest. Defaults to a 10 (for my version of sklearn)
* criterion: The function to measure the quality of a split. Supported criteria are “gini” for the Gini impurity and “entropy” for the information gain
* min_weight_fraction_leaf: The minimum weighted fraction of the sum total of weights (of all the input samples) required to be at a leaf node
* max_depth: depth of the tree
* max_features: The number of features to consider when looking for the best split

Can we further optimize this classifier?

In [ ]:
# defining the parameters to tune
param_grid = { 
    'classifier__n_estimators': [10, 100, 200],
    'classifier__max_features': ['auto', 'sqrt', 'log2'],
    'classifier__min_weight_fraction_leaf': [0],
    'classifier__max_depth' : [None,1, 2, 3],
    'classifier__criterion' :['gini', 'entropy']}

from sklearn.model_selection import GridSearchCV
CV = GridSearchCV(rf, param_grid, n_jobs= 1)
                  
CV.fit(X_train, y_train)  
print(CV.best_params_)    
print(CV.best_score_)

For me the best params were:
```
{'classifier__criterion': 'entropy', 'classifier__max_depth': None, 'classifier__max_features': 'sqrt', 'classifier__min_weight_fraction_leaf': 0, 'classifier__n_estimators': 100}
```


As you see above we managed to optimize our classifier model further (ok, just a little bit, score from 0.94 to 0.956)

# Explaining / Present Results

# Eli5

In [ ]:
!pip install -U eli5

In [37]:
import eli5

In [38]:
rfc = RandomForestClassifier(criterion='gini', max_depth=None, max_features='sqrt', n_estimators=100)
pipe = Pipeline(steps=[('preprocessor', preprocessor),
                      ('classifier',  rfc)])
    
model = pipe.fit(X_train, y_train)

In [39]:
onehot_columns = list(pipe.named_steps['preprocessor'].named_transformers_['cat'].named_steps['onehotencoder'].get_feature_names(input_features=categorical_features))
numeric_features_list = list(numeric_features)
numeric_features_list.extend(onehot_columns)

In [40]:
eli5.explain_weights(pipe.named_steps['classifier'], top=50, feature_names=numeric_features_list)

Explanation(estimator="RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',\n            max_depth=None, max_features='sqrt', max_leaf_nodes=None,\n            min_impurity_decrease=0.0, min_impurity_split=None,\n            min_samples_leaf=1, min_samples_split=2,\n            min_weight_fraction_leaf=0.0, n_estimators=100, n_jobs=None,\n            oob_score=False, random_state=None, verbose=0,\n            warm_start=False)", description='\nRandom forest feature importances; values are numbers 0 <= x <= 1;\nall values sum to 1.\n', error=None, method='feature importances', is_regression=False, targets=None, feature_importances=FeatureImportances(importances=[FeatureWeight(feature='num_lab_procedures', weight=0.05083093792141481, std=0.01179092576170172, value=None), FeatureWeight(feature='num_medications', weight=0.0458741931456085, std=0.010600432547876015, value=None), FeatureWeight(feature='time_in_hospital', weight=0.03514067869416614, std=0.008326069243246807, value=None), FeatureWeight(feature='number_inpatient', weight=0.032395092119022535, std=0.012737120040254629, value=None), FeatureWeight(feature='number_diagnoses', weight=0.031642797681133344, std=0.007992119357028843, value=None), FeatureWeight(feature='num_procedures', weight=0.02661588241818991, std=0.007147044456605648, value=None), FeatureWeight(feature='discharge_disposition_id', weight=0.02585361063097006, std=0.008024187186627693, value=None), FeatureWeight(feature='admission_type_id', weight=0.02303428192803828, std=0.007372784678835795, value=None), FeatureWeight(feature='admission_source_id', weight=0.019704868307959676, std=0.00623271327463943, value=None), FeatureWeight(feature='number_outpatient', weight=0.014236054795328102, std=0.004643189739007155, value=None), FeatureWeight(feature='age_[70-80)', weight=0.010794394051452091, std=0.004330896155357995, value=None), FeatureWeight(feature='number_emergency', weight=0.010431754842689161, std=0.004550397837910397, value=None), FeatureWeight(feature='gender_Male', weight=0.010228395797167677, std=0.003323498625930303, value=None), FeatureWeight(feature='gender_Female', weight=0.009743444764465934, std=0.004274430066086158, value=None), FeatureWeight(feature='age_[60-70)', weight=0.009112105119694918, std=0.0037071220722911698, value=None), FeatureWeight(feature='race_Caucasian', weight=0.00905165335354491, std=0.004320353293118708, value=None), FeatureWeight(feature='insulin_Steady', weight=0.009021972671883963, std=0.0038072415936328953, value=None), FeatureWeight(feature='age_[50-60)', weight=0.008780697904223336, std=0.003953834781962472, value=None), FeatureWeight(feature='age_[80-90)', weight=0.00808039793032473, std=0.0043934886635859935, value=None), FeatureWeight(feature='insulin_No', weight=0.00783386051298465, std=0.0032907392825255685, value=None), FeatureWeight(feature='change_No', weight=0.007785148617462792, std=0.00338539500976569, value=None), FeatureWeight(feature='metformin_No', weight=0.007769139580854189, std=0.0034100476924404603, value=None), FeatureWeight(feature='diag_3_250', weight=0.007715578408145863, std=0.003348334805757776, value=None), FeatureWeight(feature='diag_1_428', weight=0.0074762055561101715, std=0.0037090129927821415, value=None), FeatureWeight(feature='change_Ch', weight=0.007406091891869402, std=0.0036543380981037795, value=None), FeatureWeight(feature='A1Cresult_None', weight=0.007397821864247536, std=0.003609997593820953, value=None), FeatureWeight(feature='race_AfricanAmerican', weight=0.007311517928091315, std=0.0035401420343442926, value=None), FeatureWeight(feature='metformin_Steady', weight=0.007293777667227813, std=0.0033856437303649177, value=None), FeatureWeight(feature='diag_3_401', weight=0.006530507297535858, std=0.0027258101738564956, value=None), FeatureWeight(feature='insulin_Down', weight=0.006443274729793911, std=0.0034300100443636966, value=None), FeatureWeight(feature='glipizide_Steady', weight=0.006431126243093514, s

<div class="alert alert-danger"> 💡 <strong> Attention</strong>

Above analysis explains what features wheigh in on making a decision.  
However, It also makes paifully clear that because of the preprocessing there are about 1000 features. ( mostly because of using the OneHotEncoder ).

In my opinion `diag_1, diag_2 and diag_3` are too specifically categorized.
To be sure about this we would have to speak to a domain expert.

However, I will go ahead an reduce the specificity of the columns by grouping them in categories. This will introduce some bias. Let's see the effects!
</div>

# Data prep, again!

### diag_1, diag_2, diag_3
These indicate the category of diagnosis.  
These columns have a lot of different categories! I am going to group them according to this [list](https://nl.wikipedia.org/wiki/ICD-9) based on the first tree letters.

I suspect there will be helpfull information in the fact that a patient was correctly / incorrectly diagnosed.


Below I am executing the following steps:
* convert each of the diag columns to a numeric type
* categorize all the values according to the list of diagnosis linked above
* labeling all the categories
  * E = Endocrinic
  * C = Cardiovascular disease
  * R = Respatory system
  * I = Ill defined conditions
  * O = Other

In [41]:
df = data

In [42]:
for diag in ['diag_1', 'diag_2', 'diag_3']:
    df[diag].fillna(0, inplace=True)
    df[diag] = pd.to_numeric(df[diag], errors='coerce').fillna(0)
    df = df[pd.to_numeric(df[diag], errors='coerce').notnull()]
    df[diag] = pd.to_numeric(df[diag])
    df[diag] = df[diag].round(0).astype(int)
    
    bins = [0, 240, 280, 390, 460, 520, 780, 800, np.inf]
    names = ['O1', 'E', 'O2', 'C', 'R', 'O3', 'I', 'O4']
    df[diag] = pd.cut(df[diag], bins, labels=names).astype('object')
    
    df.loc[ df[diag] == 'O1', diag] = 'O'
    df.loc[ df[diag] == 'O2', diag] = 'O'
    df.loc[ df[diag] == 'O3', diag] = 'O'
    df.loc[ df[diag] == 'O4', diag] = 'O'
    
    print('Values in {}: '.format(diag))
    print(df[diag].value_counts())
    print('\n')

Values in diag_1: 
O    1213
C     930
E     363
R     307
I     138
Name: diag_1, dtype: int64


Values in diag_2: 
O    955
C    946
E    622
R    287
I     98
Name: diag_2, dtype: int64


Values in diag_3: 
C    885
O    840
E    807
R    177
I     96
Name: diag_3, dtype: int64




In [43]:
print(df.shape)
df.head()

(3000, 45)


race  gender      age  admission_type_id  \
0        Caucasian  Female  [50-60)                  1   
1  AfricanAmerican    Male  [40-50)                  1   
2        Caucasian    Male  [30-40)                  3   
3        Caucasian  Female  [70-80)                  3   
4        Caucasian  Female  [50-60)                  1   

   discharge_disposition_id  admission_source_id  time_in_hospital  \
0                         3                    7                 9   
1                         1                    7                 1   
2                         1                    1                 5   
3                        18                    1                 1   
4                         6                    7                 3   

   num_lab_procedures  num_procedures  num_medications  number_outpatient  \
0                  65               4               28                  0   
1                  38               1                6                  1   
2                  42               3               26                  0   
3                   1               1                4                  0   
4                  42               1                9                  0   

   number_emergency  number_inpatient diag_1 diag_2 diag_3  number_diagnoses  \
0                 0                 0      C      O      O                 9   
1                 0                 0      E      C      I                 5   
2                 0                 0      O      I      E                 4   
3                 0                 0      E      E    NaN                 2   
4                 0                 0      E      O    NaN                 9   

  max_glu_serum A1Cresult metformin repaglinide nateglinide chlorpropamide  \
0          None      None        No          No          No             No   
1          None      None    Steady          No          No             No   
2          None      None        No          No          No             No   
3          None      None        No          No          No         Steady   
4          None      Norm        No          No          No             No   

  glimepiride acetohexamide glipizide glyburide tolbutamide pioglitazone  \
0          No            No        No        No          No           No   
1          No            No        No        No          No           No   
2          No            No        No        No          No           No   
3          No            No        No        No          No           No   
4          No            No        No        No          No           No   

  rosiglitazone acarbose miglitol troglitazone tolazamide examide citoglipton  \
0            No       No       No           No         No      No          No   
1            No       No       No           No         No      No          No   
2            No       No       No           No         No      No          No   
3            No       No       No           No         No      No          No   
4            No       No       No           No         No      No          No   

  insulin glyburide-metformin glipizide-metformin glimepiride-pioglitazone  \
0    Down                  No                  No                       No   
1      No                  No                  No                       No   
2    Down                  No                  No                       No   
3      No                  No                  No                       No   
4    Down                  No                  No                       No   

  metformin-rosiglitazone metformin-pioglitazone change diabetesMed readmitted  
0                      No                     No     Ch         Yes          B  
1                      No                     No     No         Yes          B  
2                      No                     No     Ch         Yes          B  
3                      No                     No     No         Yes          A  
4            

In [44]:
data = df

# Evaluate Algoritms

## Evaluation different models
We will be looping over different classification models.

<div class="alert alert-info"> 💡 <strong> Model evalution</strong>

Model evaluation strategy:
* loop over the different classification model
* score each model on accuracy
* calculate f1 score for each model
* retain the score of each model
* compare the score of each model
</div>

[https://medium.com/vickdata/a-simple-guide-to-scikit-learn-pipelines-4ac0d974bdcf#49aa](https://medium.com/vickdata/a-simple-guide-to-scikit-learn-pipelines-4ac0d974bdcf#49aa)

In [45]:
X = df.iloc[:, df.columns != 'readmitted']
y = df.iloc[:, df.columns == 'readmitted']

In [46]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

In [47]:
target_names = ['A', 'B']
f1_scores = {}

for classifier in classifiers:
    print(type(classifier).__name__)
    
    pipe = Pipeline(steps=[('preprocessor', preprocessor),('classifier', classifier)])
    pipe.fit(X_train, y_train)   
    
    print("model accuracy score: %.3f" % pipe.score(X_test, y_test))
    
    y_pred = pipe.predict(X_test)
    report = classification_report(y_test, y_pred, target_names=target_names)
    print(report)
    
    report = classification_report(y_test, y_pred, target_names=target_names, output_dict=True)
    f1_scores[type(classifier).__name__] = report.get('weighted avg').get('f1-score')
    
    print('\n')

KNeighborsClassifier
model accuracy score: 0.546
              precision    recall  f1-score   support

           A       0.56      0.54      0.55       461
           B       0.53      0.55      0.54       439

   micro avg       0.55      0.55      0.55       900
   macro avg       0.55      0.55      0.55       900
weighted avg       0.55      0.55      0.55       900



DecisionTreeClassifier
model accuracy score: 0.502
              precision    recall  f1-score   support

           A       0.52      0.46      0.49       461
           B       0.49      0.55      0.52       439

   micro avg       0.50      0.50      0.50       900
   macro avg       0.50      0.50      0.50       900
weighted avg       0.50      0.50      0.50       900



RandomForestClassifier
model accuracy score: 0.561
              precision    recall  f1-score   support

           A       0.57      0.59      0.58       461
           B       0.55      0.54      0.54       439

   micro avg       0.56    

In [48]:
from pprint import pprint

pprint(f1_scores)

{'DecisionTreeClassifier': 0.5012624727062992,
 'KNeighborsClassifier': 0.5456032456420122,
 'LogisticRegression': 0.5875461725279033,
 'MLPClassifier': 0.5579930784822762,
 'RandomForestClassifier': 0.5608642013470303}


## Re - Evaluation different models
We will be looping over different classification models.

<div class="alert alert-info"> 💡 <strong> Re-evaluation</strong>   
!! Categorizing the diagnostic columns has not reallly improved the result for our models.  Too bad.
The performance is still pretty poor.
</div>

#### Model Tuning: again I will do model tuning for the RandomForestClassifier

# Improve Results

### Model tuning

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rf = Pipeline(steps=[('preprocessor', preprocessor),
                      ('classifier', RandomForestClassifier())])

In [ ]:
rf.fit(X_train, y_train)

In [ ]:
y_pred = rf.predict(X_test)

In [ ]:
# ignore warnings for deprecated code in the libraries that we use
warnings.filterwarnings("ignore")

We are going to tune the parameters of our model. These parameters are:
* n_estimators: actually the number of trees in our forest. Defaults to a 10 (for my version of sklearn)
* criterion: The function to measure the quality of a split. Supported criteria are “gini” for the Gini impurity and “entropy” for the information gain
* min_weight_fraction_leaf: The minimum weighted fraction of the sum total of weights (of all the input samples) required to be at a leaf node
* max_depth: depth of the tree
* max_features: The number of features to consider when looking for the best split

Can we further optimize this classifier?

In [ ]:
# defining the parameters to tune
param_grid = { 
    'classifier__n_estimators': [10, 100, 200],
    'classifier__max_features': ['auto', 'sqrt', 'log2'],
    'classifier__min_weight_fraction_leaf': [0],
    'classifier__max_depth' : [None,1, 2, 3],
    'classifier__criterion' :['gini', 'entropy']}

from sklearn.model_selection import GridSearchCV
CV = GridSearchCV(rf, param_grid, n_jobs= 1)
                  
CV.fit(X_train, y_train)  
print(CV.best_params_)    
print(CV.best_score_)

For me the best params were:
```
{'classifier__criterion': 'gini', 'classifier__max_depth': None, 'classifier__max_features': 'sqrt', 'classifier__min_weight_fraction_leaf': 0, 'classifier__n_estimators': 200}
```


### Focussing on the most important features? 
Currently our classifier scores 0.61. **Maybe we can still improve by focussing on the most important features only.**

In [49]:
rfc = RandomForestClassifier(criterion='gini', max_depth=None, max_features='sqrt', n_estimators=200)
pipe = Pipeline(steps=[('preprocessor', preprocessor),
                      ('classifier',  rfc)])
    
model = pipe.fit(X_train, y_train)

# Eli5

In [ ]:
!pip install -U eli5

In [50]:
import eli5

In [51]:
onehot_columns = list(pipe.named_steps['preprocessor'].named_transformers_['cat'].named_steps['onehotencoder'].get_feature_names(input_features=categorical_features))
numeric_features_list = list(numeric_features)
numeric_features_list.extend(onehot_columns)

In [52]:
eli5.explain_weights(pipe.named_steps['classifier'], top=50, feature_names=numeric_features_list)

Explanation(estimator="RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',\n            max_depth=None, max_features='sqrt', max_leaf_nodes=None,\n            min_impurity_decrease=0.0, min_impurity_split=None,\n            min_samples_leaf=1, min_samples_split=2,\n            min_weight_fraction_leaf=0.0, n_estimators=200, n_jobs=None,\n            oob_score=False, random_state=None, verbose=0,\n            warm_start=False)", description='\nRandom forest feature importances; values are numbers 0 <= x <= 1;\nall values sum to 1.\n', error=None, method='feature importances', is_regression=False, targets=None, feature_importances=FeatureImportances(importances=[FeatureWeight(feature='num_lab_procedures', weight=0.07881376388940374, std=0.01314748863104045, value=None), FeatureWeight(feature='num_medications', weight=0.07478692887246707, std=0.014280817443587559, value=None), FeatureWeight(feature='time_in_hospital', weight=0.05408577321644459, std=0.0108426511056035, value=None), FeatureWeight(feature='number_diagnoses', weight=0.04580710366111444, std=0.009639745177240894, value=None), FeatureWeight(feature='number_inpatient', weight=0.04137708842364688, std=0.011598209429343642, value=None), FeatureWeight(feature='num_procedures', weight=0.038729003789434106, std=0.008667172847549035, value=None), FeatureWeight(feature='discharge_disposition_id', weight=0.03622418781424361, std=0.008747441101121719, value=None), FeatureWeight(feature='admission_type_id', weight=0.03376631543790834, std=0.008846106705436636, value=None), FeatureWeight(feature='admission_source_id', weight=0.028455044227948378, std=0.00757853604768366, value=None), FeatureWeight(feature='number_outpatient', weight=0.01933519475568388, std=0.006161923787046728, value=None), FeatureWeight(feature='diag_3_C', weight=0.015755902332446982, std=0.005925278684867881, value=None), FeatureWeight(feature='diag_1_O', weight=0.015086647001059937, std=0.005720133082277931, value=None), FeatureWeight(feature='diag_2_O', weight=0.015012706693419548, std=0.004998738861708255, value=None), FeatureWeight(feature='gender_Male', weight=0.014706446774652906, std=0.005591683727355528, value=None), FeatureWeight(feature='age_[70-80)', weight=0.014341662531411257, std=0.005065514118190234, value=None), FeatureWeight(feature='diag_3_E', weight=0.014185334077938084, std=0.005039211746532759, value=None), FeatureWeight(feature='diag_2_C', weight=0.014150803057688318, std=0.005214180762417953, value=None), FeatureWeight(feature='diag_1_C', weight=0.014095145076439974, std=0.005868083283736968, value=None), FeatureWeight(feature='number_emergency', weight=0.014090824491455831, std=0.00551967991652194, value=None), FeatureWeight(feature='gender_Female', weight=0.01405240824653934, std=0.0047871490322513, value=None), FeatureWeight(feature='diag_3_O', weight=0.013870400554249762, std=0.005254875466249003, value=None), FeatureWeight(feature='age_[60-70)', weight=0.01331615041873965, std=0.005232451569113799, value=None), FeatureWeight(feature='race_Caucasian', weight=0.013197216570890917, std=0.005600457196481559, value=None), FeatureWeight(feature='diag_2_E', weight=0.0130474559666685, std=0.005140376173226062, value=None), FeatureWeight(feature='insulin_Steady', weight=0.01256619592677373, std=0.005412896812449345, value=None), FeatureWeight(feature='age_[80-90)', weight=0.012563078712396886, std=0.004737974102015115, value=None), FeatureWeight(feature='age_[50-60)', weight=0.011979817467911855, std=0.004646489884135004, value=None), FeatureWeight(feature='insulin_No', weight=0.011553478491502069, std=0.004891014553808731, value=None), FeatureWeight(feature='change_Ch', weight=0.010617388539973505, std=0.005225267466164041, value=None), FeatureWeight(feature='A1Cresult_None', weight=0.010610802739860512, std=0.004389519156588556, value=None), FeatureWeight(feature='metformin_No', weight=0.0101782335183873, std=0.004564647606716187, value=None), FeatureWeight(feature=

*Yes! After categorizing the diagnostic columns (diag_1, diag_2, diag_3)
we can see that being wrongly diagnosed the first or second time actually has a pretty high effect on chances of readdmittance!

Let's only keep the 15 most important feature and build a new model

In [53]:
df = data

In [54]:
df = df[['num_lab_procedures', 'num_medications', 'time_in_hospital', 'number_diagnoses', 'number_inpatient',
           'num_procedures', 'discharge_disposition_id', 'admission_type_id', 'admission_source_id', 'number_outpatient',
           'diag_1', 'diag_2', 'diag_3', 'readmitted']].copy()
print(df.shape)
df.head()

(3000, 14)


num_lab_procedures  num_medications  time_in_hospital  number_diagnoses  \
0                  65               28                 9                 9   
1                  38                6                 1                 5   
2                  42               26                 5                 4   
3                   1                4                 1                 2   
4                  42                9                 3                 9   

   number_inpatient  num_procedures  discharge_disposition_id  \
0                 0               4                         3   
1                 0               1                         1   
2                 0               3                         1   
3                 0               1                        18   
4                 0               1                         6   

   admission_type_id  admission_source_id  number_outpatient diag_1 diag_2  \
0                  1                    7                  0      C      O   
1                  1                    7                  1      E      C   
2                  3                    1                  0      O      I   
3                  3                    1                  0      E      E   
4                  1                    7                  0      E      O   

  diag_3 readmitted  
0      O          B  
1      I          B  
2      E          B  
3    NaN          A  
4    NaN          A

In [55]:
X = df.iloc[:, df.columns != 'readmitted']
y = df.iloc[:, df.columns == 'readmitted']

In [56]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

In [57]:
X_train.head()

num_lab_procedures  num_medications  time_in_hospital  number_diagnoses  \
1732                   9                4                 3                 7   
2440                  51               15                 2                 6   
1232                  41               12                 2                 5   
1081                  58                5                 2                 5   
2920                  54               57                11                 5   

      number_inpatient  num_procedures  discharge_disposition_id  \
1732                 1               0                         1   
2440                 0               2                         1   
1232                 0               0                         1   
1081                 1               0                         1   
2920                 0               3                         1   

      admission_type_id  admission_source_id  number_outpatient diag_1 diag_2  \
1732                  1                    2                  0      C      E   
2440                  1                    7                  0      C      C   
1232                  2                    1                  0      C      O   
1081                  1                    7                  1      E      E   
2920                  1                    7                  0      C      C   

     diag_3  
1732      C  
2440      C  
1232      E  
1081      E  
2920      E

In [58]:
y_train.head()

readmitted
1732          A
2440          B
1232          B
1081          B
2920          A

# Improve Results

### Model tuning

In [59]:
numeric_features = X_train.select_dtypes(include=['int64', 'float64']).columns
print(numeric_features)

Index(['num_lab_procedures', 'num_medications', 'time_in_hospital',
       'number_diagnoses', 'number_inpatient', 'num_procedures',
       'discharge_disposition_id', 'admission_type_id', 'admission_source_id',
       'number_outpatient'],
      dtype='object')


In [60]:
categorical_features = X_train.select_dtypes(include=['object']).columns
print(categorical_features)

Index(['diag_1', 'diag_2', 'diag_3'], dtype='object')


In [61]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder

numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
    ('scaler', StandardScaler())])

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='incomplete')),
    ('onehotencoder', OneHotEncoder(handle_unknown='ignore'))])

In [62]:
from sklearn.compose import ColumnTransformer

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)])

In [63]:
from sklearn.ensemble import RandomForestClassifier
rf = Pipeline(steps=[('preprocessor', preprocessor),
                      ('classifier', RandomForestClassifier())])

In [64]:
rf.fit(X_train, y_train)

Pipeline(memory=None,
     steps=[('preprocessor', ColumnTransformer(n_jobs=None, remainder='drop', sparse_threshold=0.3,
         transformer_weights=None,
         transformers=[('num', Pipeline(memory=None,
     steps=[('imputer', SimpleImputer(copy=True, fill_value=None, missing_values=nan, strategy='mean',
       verbose...obs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False))])

In [65]:
y_pred = rf.predict(X_test)

In [66]:
# ignore warnings for deprecated code in the libraries that we use
warnings.filterwarnings("ignore")

We are going to tune the parameters of our model. These parameters are:
* n_estimators: actually the number of trees in our forest. Defaults to a 10 (for my version of sklearn)
* criterion: The function to measure the quality of a split. Supported criteria are “gini” for the Gini impurity and “entropy” for the information gain
* min_weight_fraction_leaf: The minimum weighted fraction of the sum total of weights (of all the input samples) required to be at a leaf node
* max_depth: depth of the tree
* max_features: The number of features to consider when looking for the best split

Can we further optimize this classifier?

In [67]:
# defining the parameters to tune
param_grid = { 
    'classifier__n_estimators': [10, 100, 200],
    'classifier__max_features': ['auto', 'sqrt', 'log2'],
    'classifier__min_weight_fraction_leaf': [0],
    'classifier__max_depth' : [None,1, 2, 3],
    'classifier__criterion' :['gini', 'entropy']}

from sklearn.model_selection import GridSearchCV
CV = GridSearchCV(rf, param_grid, n_jobs= 1)
                  
CV.fit(X_train, y_train)  
print(CV.best_params_)    
print(CV.best_score_)

{'classifier__criterion': 'entropy', 'classifier__max_depth': 3, 'classifier__max_features': 'sqrt', 'classifier__min_weight_fraction_leaf': 0, 'classifier__n_estimators': 10}
0.6195238095238095


This improved the model a little bit.  It did simplify the model though. I decide to keep it this way!
That is understandable since the other features had a weight variance that was larger then their actual weight.
This actually made them behave randomly.


# Explaining / Present Results

In [81]:
rfc = RandomForestClassifier(criterion='gini', max_depth=3, max_features='sqrt', n_estimators=10)
pipe = Pipeline(steps=[('preprocessor', preprocessor),
                      ('classifier',  rfc)])
    
model = pipe.fit(X_train, y_train)

## Eli5

In [82]:
import eli5

In [83]:
onehot_columns = list(pipe.named_steps['preprocessor'].named_transformers_['cat'].named_steps['onehotencoder'].get_feature_names(input_features=categorical_features))
numeric_features_list = list(numeric_features)
numeric_features_list.extend(onehot_columns)

In [79]:
eli5.explain_weights(pipe.named_steps['classifier'], top=50, feature_names=numeric_features_list)

Explanation(estimator="RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',\n            max_depth=None, max_features='sqrt', max_leaf_nodes=None,\n            min_impurity_decrease=0.0, min_impurity_split=None,\n            min_samples_leaf=1, min_samples_split=2,\n            min_weight_fraction_leaf=0.0, n_estimators=200, n_jobs=None,\n            oob_score=False, random_state=None, verbose=0,\n            warm_start=False)", description='\nRandom forest feature importances; values are numbers 0 <= x <= 1;\nall values sum to 1.\n', error=None, method='feature importances', is_regression=False, targets=None, feature_importances=FeatureImportances(importances=[FeatureWeight(feature='num_lab_procedures', weight=0.14887221493069727, std=0.0174027351073034, value=None), FeatureWeight(feature='num_medications', weight=0.1358584203734282, std=0.015064699173014378, value=None), FeatureWeight(feature='time_in_hospital', weight=0.09408890737544491, std=0.014789303342249562, value=None), FeatureWeight(feature='number_diagnoses', weight=0.07083123801938837, std=0.011089943845031053, value=None), FeatureWeight(feature='num_procedures', weight=0.06459198475830658, std=0.010603833841274377, value=None), FeatureWeight(feature='discharge_disposition_id', weight=0.060219740160710106, std=0.009893915541147089, value=None), FeatureWeight(feature='number_inpatient', weight=0.053387252761416724, std=0.009705806148806308, value=None), FeatureWeight(feature='admission_type_id', weight=0.05066220973874424, std=0.009434567450179936, value=None), FeatureWeight(feature='admission_source_id', weight=0.042177360162421816, std=0.009422625671612065, value=None), FeatureWeight(feature='number_outpatient', weight=0.031040953209615026, std=0.00786371320684903, value=None), FeatureWeight(feature='diag_3_C', weight=0.02204864177699195, std=0.006744173270527104, value=None), FeatureWeight(feature='diag_3_E', weight=0.021750126508590824, std=0.006212828762998827, value=None), FeatureWeight(feature='diag_2_C', weight=0.021350514813738015, std=0.007340527561175947, value=None), FeatureWeight(feature='diag_3_O', weight=0.020698926136502246, std=0.00610715589762167, value=None), FeatureWeight(feature='diag_2_O', weight=0.020294724380184283, std=0.00611545011089645, value=None), FeatureWeight(feature='diag_2_E', weight=0.01876597162943218, std=0.005997186304279693, value=None), FeatureWeight(feature='diag_1_O', weight=0.01806488155740335, std=0.006400659099871558, value=None), FeatureWeight(feature='diag_1_C', weight=0.017597044455411607, std=0.00669999389480993, value=None), FeatureWeight(feature='diag_2_R', weight=0.01380264086919807, std=0.0044075926805924315, value=None), FeatureWeight(feature='diag_1_E', weight=0.012194327436156385, std=0.004814787253622043, value=None), FeatureWeight(feature='diag_1_R', weight=0.010501796239207628, std=0.004280443286797356, value=None), FeatureWeight(feature='diag_3_R', weight=0.01022736001369772, std=0.004088974208139946, value=None), FeatureWeight(feature='diag_3_incomplete', weight=0.009981808776318234, std=0.004434203103131178, value=None), FeatureWeight(feature='diag_2_I', weight=0.007573802568320495, std=0.003621051729620297, value=None), FeatureWeight(feature='diag_1_I', weight=0.007242084170725879, std=0.0033876543495013167, value=None), FeatureWeight(feature='diag_3_I', weight=0.007111646047788891, std=0.0032325299392794997, value=None), FeatureWeight(feature='diag_2_incomplete', weight=0.006594121740594314, std=0.003565482741166689, value=None), FeatureWeight(feature='diag_1_incomplete', weight=0.0024692993895647566, std=0.001975652588229972, value=None)], remaining=0), decision_tree=None, highlight_spaces=None, transition_features=None, image=None)

In [80]:
target_names = ['A', 'B']
print(type(rfc).__name__)

pipe = Pipeline(steps=[('preprocessor', preprocessor),('classifier', rfc)])
pipe.fit(X_train, y_train)   

print("model accuracy score: %.3f" % pipe.score(X_test, y_test))

y_pred = pipe.predict(X_test)
report = classification_report(y_test, y_pred, target_names=target_names)
print(report)

report = classification_report(y_test, y_pred, target_names=target_names, output_dict=True)



RandomForestClassifier
model accuracy score: 0.567
              precision    recall  f1-score   support

           A       0.58      0.53      0.56       461
           B       0.55      0.60      0.58       439

   micro avg       0.57      0.57      0.57       900
   macro avg       0.57      0.57      0.57       900
weighted avg       0.57      0.57      0.57       900



# Conclusions

<div class="alert alert-info"> 💡 <strong>Conclusions</strong>  

* Problem was attacked as a binary classifiction problem where we try to predict wether a patient will be readmitted or not  
* Unnec. columns were dropped  
* Multiple models were evaluated  
* The RandomForestClassifier was optimized  
* We used a pipeline for preprocessing and fitting steps  
* We discovered feature importance. We found that the model could be simplified by using only the 15 most important features:
        `['num_lab_procedures', 'num_medications', 'time_in_hospital', 'number_diagnoses', 'number_inpatient',
           'num_procedures', 'discharge_disposition_id', 'admission_type_id', 'admission_source_id', 'number_outpatient',
           'diag_1', 'diag_2', 'diag_3', 'readmitted']`
* Simplified model: RFC scores are as follows:  
model accuracy score: 0.598  
                  precision    recall  f1-score   support

               A       0.61      0.59      0.60       457
               B       0.59      0.61      0.60       443

       micro avg       0.60      0.60      0.60       900
       macro avg       0.60      0.60      0.60       900
       weighted avg    0.60      0.60      0.60       900
</div>

